### PR TITLE
Add NPM workbook for AKS

### PR DESCRIPTION
**Pull request by huntergregory**
Adds a workbook for customers using Network Policy Manager (NPM) to help them understand their configuration. 

# Workbook Screenshots
image

image

image

# Setup & Messing Around with the Workbook
## NPM
AKS isn't yet configured with the NPM changes that provide the Prometheus metrics. 
Apply the following file to your cluster to enable the right NPM (`kubectl apply -f azure-npm.yaml`): 
```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: azure-npm
  namespace: kube-system
  labels:
    addonmanager.kubernetes.io/mode: EnsureExists
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: azure-npm
  namespace: kube-system
  labels:
    addonmanager.kubernetes.io/mode: EnsureExists
rules:
  - apiGroups:
    - ""
    resources:
      - pods
      - nodes
      - namespaces
    verbs:
      - get
      - list
      - watch
  - apiGroups:
    - networking.k8s.io
    resources:
      - networkpolicies
    verbs:
      - get
      - list
      - watch
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: azure-npm-binding
  namespace: kube-system
  labels:
    addonmanager.kubernetes.io/mode: EnsureExists
subjects:
  - kind: ServiceAccount
    name: azure-npm
    namespace: kube-system
roleRef:
  kind: ClusterRole
  name: azure-npm
  apiGroup: rbac.authorization.k8s.io
---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: azure-npm
  namespace: kube-system
  labels:
    app: azure-npm
    addonmanager.kubernetes.io/mode: EnsureExists
spec:
  selector:
    matchLabels:
      k8s-app: azure-npm
  template:
    metadata:
      labels:
        k8s-app: azure-npm
      annotations:
        scheduler.alpha.kubernetes.io/critical-pod: ''
        prometheus.io/scrape: "true"
        prometheus.io/port:   "10091"
    spec:
      priorityClassName: system-node-critical
      tolerations:
      - operator: "Exists"
        effect: NoExecute
      - operator: "Exists"
        effect: NoSchedule
      - key: CriticalAddonsOnly
        operator: Exists
      nodeSelector:
        beta.kubernetes.io/os: linux
        kubernetes.io/role: agent
      containers:
        - name: azure-npm
          image: thugreg/azure-npm:2.1
          securityContext:
            privileged: true
          env:
            - name: HOSTNAME
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: spec.nodeName
          volumeMounts:
          - name: xtables-lock
            mountPath: /run/xtables.lock
          - name: log
            mountPath: /var/log
          - name: protocols
            mountPath: /etc/protocols
      hostNetwork: true
      volumes:
      - name: log
        hostPath:
          path: /var/log
          type: Directory
      - name: xtables-lock
        hostPath:
          path: /run/xtables.lock
          type: File
      - name: protocols
        hostPath:
          path: /etc/protocols
          type: File
      serviceAccountName: azure-npm
---
apiVersion: v1
kind: Service
metadata:
  name: npm-metrics-cluster-service
  namespace: kube-system
  labels:
    app: npm-metrics
spec:
  selector:
    k8s-app: azure-npm
  ports:
    - port: 9000
      targetPort: 10091
```

### Cluster Level Scrape
kubernetes_services = ["http://npm-metrics-cluster-service.kube-system:9000/metrics"]
fieldpass = ["npm_num_policies", "npm_num_iptables_rules", "npm_num_ipsets", "num_ipset_entries”]

### Node Level Scrapes
urls = ["http://$NODE_IP:10091/metrics"]
fieldpass = ["npm_add_policy_exec_time", "npm_add_iptables_rule_exec_time", "npm_add_ipset_exec_time"]

### Advanced Metrics
Change in Cluster Level Scrape
fieldpass = ["npm_num_policies", "npm_num_iptables_rules", "npm_num_ipsets", "num_ipset_entries", **"npm_ipset_counts"**]

### Production Notes
Eventually Vishwanath and I were planning to have booleans in the agentconfig.yaml where customers can include 
```
npm_metrics = true
advanced_npm_metrics = true
```

### CODE REVIEW

**Comment by BLuEScioN** [comment]
Can you adjust the widths of the controls on the IPSet Entries Tab. Try to eliminate dead white space. Something like this:

image

Also, have you tested this in prod with feature flag? Let's do that first. We can do that by adding "isPreview": true to the settings.json

**Comment by huntergregory** [comment]
adding Neha and Shufang @csfmomo and @neaggarwMS

**Comment by huntergregory** [comment]
@BLuEScioN do you think the extra space in the grid (on the right) looks bad? 
image

File: Workbooks/AKS/NPM Configuration/NPM Configuration.workbook [hunk 1]
```diff
@@ -0,0 +1,604 @@
...
+              }
+            },
+            "customWidth": "15",
+            "name": "Total IPSet Entries"
+          },
+          {
+            "type": 3,
+            "content": {
+              "version": "KqlItem/1.0",
+              "query": "InsightsMetrics\r\n| where Namespace == \"prometheus\"\r\n| where Name == \"npm_ipset_counts\"\r\n| where TimeGenerated > ago(2m)\r\n| extend tags = parse_json(Tags)\r\n| extend setName = tostring(tags.set_name)\r\n| summarize arg_max(TimeGenerated, Val) by setName\r\n| where Val > 0\r\n| order by Val desc, setName asc\r\n| project Set = setName, Entries = Val",
```
Comment (huntergregory) [inline Workbooks/AKS/NPM Configuration/NPM Configuration.workbook:364]: @vishiy this query depends upon the scrape interval for the advanced metric with high cardinality. Should I increase the time in the "ago" clause? Our NPM deletes the metrics with 0 counts, so if we don't include an ago clause, the grid will incorrectly display the last-scraped old counts for any deleted ipset.

Comment (vishiy) [reply Workbooks/AKS/NPM Configuration/NPM Configuration.workbook:364]: We will scrape all with a 1 min frequency. But rather then hard coding time, you should pass in as a parameter (thru a time selector in the workbook). Just take a look at our other workbooks in a cluster.

**Pull request merged**